[BUG] Spark 4.1: UPDATE/MERGE/DELETE fail with NoClassDefFoundError

**Describe the problem you faced**

On a Spark 4.1 runtime, any Hudi read-modify write — `UPDATE`, `MERGE INTO`, `DELETE` (the upsert path that reads existing base files) — aborts with:

```
java.lang.NoClassDefFoundError: org/apache/parquet/variant/VariantConverters$ParentConverter
Caused by: java.lang.ClassNotFoundException: org.apache.parquet.variant.VariantConverters$ParentConverter
```

A fresh `INSERT` / bulk write succeeds; only the read-modify operations fail.

This was surfaced by the custom-type hive-sync E2E harness `ITTestCustomTypeHiveSync` (introduced in #19203), specifically the Spark 4.1 matrix row added in #19216. Because the failed writes never commit, the new partition the `MERGE` should have created is missing, so the test trips on a downstream Hive `SHOW PARTITIONS` assertion — but the real failure is the `NoClassDefFoundError` above (the fixtures' unconditional `*_SUCCESS` prints mask it).

**Root cause**

> Correction: an earlier version of this issue attributed this to the bundle shipping Parquet 1.13.1. That is not the case — the Parquet versions are correct. The real cause is a shade/packaging gap, described below.

The Spark 4.1 profile uses Parquet **1.16.0** (matching Spark 4.1.1; the Spark 4.0 profile uses 1.15.2, matching Spark 4.0.2). The problem is packaging:

- Parquet **1.16.0** split VARIANT support into a **separate `org.apache.parquet:parquet-variant` module** — it does not exist for 1.15.2 / 1.13.1. `org.apache.parquet.variant.VariantConverters` lives there.
- `parquet-avro:1.16.0` — which `hoodie-spark-bundle` shades and uses on the read path — has a direct compile dependency on `parquet-variant` and references `VariantConverters`.
- The bundle's `maven-shade-plugin` `<artifactSet><includes>` is an allowlist that lists `parquet-avro` but **not** `parquet-variant`. So even though `parquet-variant` is on the (transitive) dependency tree, its classes are dropped from the bundle — and they're not on the Spark runtime classpath either → `NoClassDefFoundError`.

**To Reproduce**

1. Spark 4.1 runtime with the Hudi Spark 4.1 bundle on the classpath.
2. Create a partitioned COW table and `INSERT` rows — succeeds.
3. Run any `UPDATE` / `MERGE INTO` / `DELETE` — aborts with the `NoClassDefFoundError`.

**Expected behavior**

UPDATE/MERGE/DELETE work on Spark 4.1, as they do on Spark 3.5 / 4.0.

**Environment Description**

- Hudi version: 1.3.0-SNAPSHOT (master)
- Spark version: 4.1.1 (Parquet 1.16.0)
- Table type: COPY_ON_WRITE
- Running on Docker?: yes (integ E2E)

**Stacktrace (Spark 4.1 E2E, `ITTestCustomTypeHiveSync`)**

```
ERROR BaseSparkCommitActionExecutor: Error upserting bucketType UPDATE for partition :0
java.lang.NoClassDefFoundError: org/apache/parquet/variant/VariantConverters$ParentConverter
Caused by: java.lang.ClassNotFoundException: org.apache.parquet.variant.VariantConverters$ParentConverter
...
ERROR TaskSetManager: Task 0 in stage 41.0 failed 1 times; aborting job
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 41.0 failed 1 times, most recent failure: Lost task 0.0 in stage 41.0 (TID 54) (adhoc-1 executor driver): org.apache.hudi.exception.HoodieUpsertException: Error upserting bucketType UPDATE for partition :0
...
org.apache.hudi.exception.HoodieUpsertException: Failed to upsert for commit time 20260709102914322
```

Downstream symptom — hive-sync then correctly finds nothing new (the write never committed):

```
INFO HiveSyncTool: Last commit time synced was found to be 20260709102854659, last commit completion time is found to be 20260709102908002
INFO HoodieSyncClient: Last commit time synced is 20260709102854659, Getting commits since then
INFO HiveSyncTool: Storage partitions scan complete. Found 0
INFO HiveSyncTool: Sync complete for vector_test
```

**Proposed fix**

Add `org.apache.parquet:parquet-variant` to the `hoodie-spark-bundle` shade include list. It is already a transitive compile dependency of `parquet-avro` for Parquet 1.16.0+, so the include shades it for Spark 4.1/4.2 and is a no-op for older Spark (the artifact does not exist there). Verified: the rebuilt Spark 4.1 bundle then contains `VariantConverters$ParentConverter`.

Fix: #19235

--- a/pom.xml
+++ b/pom.xml
@@ -484,6 +484,9 @@
               <!-- afterburner module for jackson performance -->
               <include>com.fasterxml.jackson.module:jackson-module-afterburner</include>
               <include>com.fasterxml.jackson.module:jackson-module-scala_${scala.binary.version}</include>
+              <!-- Parquet 1.16.0+ (Spark 4.1+) splits VariantConverters into parquet-variant,
+                   which parquet-avro needs at runtime. No-op for older Spark. See apache/hudi#19234. -->
+              <include>org.apache.parquet:parquet-variant</include>
             </includes>
           </artifactSet>
           <relocations>

